### DISCUSSION TITLE
Sleep Hook

### DISCUSSION THREAD

**Discussion by tleish** [socketry/async#306] [Q&A]
In using Falcon, does async include a hook to monkey-patch `Kernel#sleep` with `Async#sleep`?

[em-synchrony](https://github.com/igrigorik/em-synchrony?tab=readme-ov-file#hooks) has the following hook/logic you can add to your code if using something like Goliath.

```ruby
# Redirects to EM::Synchrony.sleep
require "em-synchrony"
EM::Synchrony.on_sleep do |*args|
  EM::Synchrony.sleep(*args)
end
```

- [EM::Synchrony#on_sleep](https://github.com/igrigorik/em-synchrony/blob/99aa05b337ed1fca3759e85e71c4e928d3e3ff7c/lib/em-synchrony.rb#L139)
- [Kernel#em_synchrony_sleep_hook](https://github.com/igrigorik/em-synchrony/blob/99aa05b337ed1fca3759e85e71c4e928d3e3ff7c/lib/em-synchrony/kernel.rb)

**Comment by ioquatix** [comment]
On Ruby 3.1+ and Async 2+, `Kernel::sleep` is redirected to the event loop

CRuby: https://github.com/ruby/ruby/blob/fb4df44d1670e9d25aef6b235a7281199a177edb/process.c#L5162-L5170

Async: https://github.com/socketry/async/blob/2c5f92ed2027e50dbfa9ce599505498843c61687/lib/async/scheduler.rb#L151-L158

**Comment by tleish** [comment]
Is there a way to do this with older versions of Ruby?  (We a have not upgraded to Ruby 3 yet).

**Reply by ioquatix** [reply]
Yes, the compatible method is:

```ruby
Async do |task|
  task.sleep(n)
end
```

or

```ruby
if task = Async::Task.current?
  task.sleep(n)
else
  sleep(n)
end
```
